### PR TITLE
Ethereum analyzer

### LINKED ISSUE #266: Ethereum Analyzer
Once #262 is done, use the platform to implement the Ethereum Metadata Provider. It should be able to provide the following data for Ethereum transactions:

- recipient address (to)
~~- tokens being transferred (including ETH **and** ERC-20 tokens)~~ (tracked in separate issue)

This already will be enough for users to write "allowlist" intents, where a transaction can only send tokens to a specific address or can only call a specific Ethereum smart contract (e.g. uniswap).

Once this is done we can remove all the Ethereum-specific code from wardend.

### PR DESCRIPTION
**Pull request by pitasi**
Replace the current SignMethod.Ethereum with an analyzer (CosmWasm contract). The code cleanup in wardend will come in a later PR, for now I'm only adding the contract and fixing SpaceWard to use it.

This initial analyzer version only returns the hash of the unsigned transaction. It's easily extendable to also return some information about the transaction itself.

The address of the contract must be set in the env variables of SpaceWard (we could consider letting users decide which contract they want to use) or sending ethereum transactions will be disabled.

depends on #279 

closes #266 

## Summary by CodeRabbit

- **New Features**
  - Introduced an analyzers smart contract for 3rd party builders to provide metadata during new signature requests.
  - Added initial version of the Ethereum analyzer for handling Ethereum unsigned transactions.

- **Improvements**
  - Updated `SendEth` and `MetaMaskRequests` components to use `useEthereumTx` for improved Ethereum transaction handling.
  - Enhanced transaction broadcasting logic in `useClient` hook for better error handling and status updates.

- **Environment**
  - Added `VITE_WARDEN_ETHEREUM_ANALYZER_CONTRACT` environment variable.

- **Dependencies**
  - Updated `axios` package from `0.21.4` to `0.28.0` with new dependencies.

- **UI/UX**
  - Added `SignatureRequestDialog` component in `WalletConnect` for displaying signature request dialogs.

### CODE REVIEW

File: contracts/contracts/ethereum-analyzer/src/msg.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+use cosmwasm_schema::{cw_serde, QueryResponses};
+use cosmwasm_std::Binary;
+
+#[cw_serde]
+pub enum ExecuteMsg {
+    Analyze { input: Binary },
+}
+
+#[cw_serde]
+pub struct AnalyzeResult<T> {
```
Comment (artur-abliazimov) [inline contracts/contracts/ethereum-analyzer/src/msg.rs:10]: Can we create a separate crate with the shared structs and functions to use among contracts?

File: spaceward/src/hooks/useClient.ts [hunk 2]
```diff
@@ -90,34 +90,37 @@ export function useTx() {
...
-							id,
-							title: TxStatus.Failed,
-							description: res?.rawLog ?? 'An unexpected error has occured',
-							duration: 10000,
-						});
-					}
-				})
-				.catch((err) => {
+			try {
+				const res = await client.broadcastTx(Uint8Array.from(txRaw.encode(signed).finish()));
```
Comment (artur-abliazimov) [inline spaceward/src/hooks/useClient.ts:94]: Do we support retry logic when broadcasting fails due to the network issues? If not, does it make sense to add it?

File: spaceward/src/hooks/useEthereumTx.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import Long from "long";
+import { ethers } from "ethers";
+import useRequestSignature from "./useRequestSignature";
+import { env } from "@/env";
+
+export function useEthereumTx() {
+	const { state, error, requestSignature, reset } = useRequestSignature();
+	return {
```
Comment (artur-abliazimov) [inline spaceward/src/hooks/useEthereumTx.tsx]: A minor one: would this be more readable if we split the function declarations from the return statement? Like you did in the other places in code.

```
export function useEthereumTx() {
	const { state, error, requestSignature, reset } = useRequestSignature();

	const signRaw = async (keyId: Long, input: Uint8Array) => {
		return await requestSignature(keyId, [], input, undefined);
	};

	const signEthereumTx = async (keyId: Long, tx: ethers.Transaction) => {
		if (!env.ethereumAnalyzerContract) {
			console.warn(
				"Missing ethereumAnalyzerContract. Can't use Ethereum transactions.",
			);
			return;
		}

		const signature = await requestSignature(
			keyId,
			[env.ethereumAnalyzerContract],
			ethers.getBytes(tx.unsignedSerialized),
			undefined,
		);
		if (!signature) {
			return;
		}

		// add the signature to the transaction
		const signedTx = tx.clone();
		signedTx.signature = ethers.hexlify(signature);

		return signedTx;
	};

	return {
		state,
		error,
		reset,
		signRaw,
		signEthereumTx,
	};
}
```

File: warden/x/warden/keeper/msg_server_new_signature_request.go [hunk 4]
```diff
@@ -38,14 +39,44 @@ func (k msgServer) NewSignatureRequest(goCtx context.Context, msg *types.MsgNewS
...
+		if err != nil {
+			return nil, nil, err
+		}
+		if dfs != nil && dataForSigning != nil {
+			return nil, nil, fmt.Errorf("two or more contracts tried to set DataForSigning. Only one analyzer contract can return DataForSigning")
+		}
+		if dfs != nil {
+			dataForSigning = dfs
+		}
+		analyzerVals[contract] = vals
```
Comment (artur-abliazimov) [inline warden/x/warden/keeper/msg_server_new_signature_request.go]: Do we expect `vals` to be non-empty? If so, let's add validation for it.

File: warden/x/warden/keeper/shield.go [hunk 5]
```diff
@@ -16,33 +17,77 @@ type WardenShieldExpander struct {
...
 
 func (w WardenShieldExpander) Expand(goCtx context.Context, ident *ast.Identifier) (*ast.Expression, error) {
+	if strings.HasPrefix(ident.Value, "analyzer.") {
+		return w.expandAnalyzers(goCtx, ident)
+	}
+
 	if ident.Value == "space.owners" {
-		ctx := cosmoshield.UnwrapContext(goCtx)
-		msg := ctx.Msg()
+		return w.expandSpaceOwners(goCtx)
```
Comment (artur-abliazimov) [inline warden/x/warden/keeper/shield.go]: Formatting: let's remove the extra space.

File: warden/x/warden/keeper/shield.go [hunk 6]
```diff
@@ -16,33 +17,77 @@ type WardenShieldExpander struct {
...
+	return ast.NewArrayLiteral(&ast.ArrayLiteral{
+		Elements: owners,
+	}), nil
+}
+
+func (w WardenShieldExpander) expandAnalyzers(ctx context.Context, ident *ast.Identifier) (*ast.Expression, error) {
+	analyzerVals := analyzerValues(ctx)
+
+	ps := strings.SplitN(ident.Value, ".", 3)
+	contract := ps[1]
```
Comment (artur-abliazimov) [inline warden/x/warden/keeper/shield.go]: Should we check that `length == 3` before accessing the indexes 1 and 2?

File: warden/x/warden/keeper/msg_server_new_signature_request.go (hunk 4 above)
Comment (pitasi) [reply warden/x/warden/keeper/msg_server_new_signature_request.go]: it can be empty

File: spaceward/src/hooks/useClient.ts (hunk 2 above)
Comment (pitasi) [reply spaceward/src/hooks/useClient.ts:94]: not sure, this is just a refactoring for moving the code from promises to async/await

if there's a retry logic, is inside the client

File: warden/x/warden/keeper/shield.go (hunk 5 above)
Comment (pitasi) [reply warden/x/warden/keeper/shield.go]: done

File: warden/x/warden/keeper/shield.go (hunk 6 above)
Comment (pitasi) [reply warden/x/warden/keeper/shield.go]: you're right, done

File: contracts/contracts/ethereum-analyzer/src/msg.rs (hunk 1 above)
Comment (pitasi) [reply contracts/contracts/ethereum-analyzer/src/msg.rs:10]: yeah it makes sense to have it in a shared crate but for now it's only used there so I didn't want to add that overhead to this PR which already introduce some more important concepts

in a separate PR sure, we should move common definition in a shared crate, feel free to open an issue to track that

Comment (artur-abliazimov) [reply contracts/contracts/ethereum-analyzer/src/msg.rs:10]: Sure ) https://github.com/orgs/warden-protocol/projects/12/views/1?pane=issue&itemId=64056128

File: spaceward/src/hooks/useEthereumTx.tsx (hunk 3 above)
Comment (pitasi) [reply spaceward/src/hooks/useEthereumTx.tsx]: done

**Pull request merged**